bundles/org.eclipse.e4.emf.xpath/src/org/eclipse/e4/emf/xpath/internal/java/JavaXPathContextFactoryImpl.java

```diff
@@ -0,0 +1,271 @@
+			}
+			return first;
+		}
+
+		private XPathFunction resolveEMFFunctions(QName functionName, int arity) {
+			if (arity == 1 && EcorePackage.eNS_URI.equals(functionName.getNamespaceURI())
+					&& "eClassName".equals(functionName.getLocalPart())) {
+				return args -> {
+					Node item = getSingleNodeArgument(args);
+					EObject eObject = (EObject) EObjectContext.this.domMapping.getValue(item);
+					return eObject == null ? null : eObject.eClass().getName();
+				};
+			}
+			return null;
+		}
+
+		private static Node getSingleNodeArgument(List<?> args) throws XPathFunctionException {
+			if (args != null && args.size() == 1) {
+				Object argument = args.get(0);
+				if (argument instanceof NodeList nodeList && nodeList.getLength() == 1) {
+					return nodeList.item(0);
+				} else if (argument instanceof Node node) {
+					return node;
+				}
+			}
+			throw new XPathFunctionException("Not a single node list: " + args);
+		}
+	}
+
+	private static Element createElement(EObject eObject, Document document, DOMMapping domMapper) {
+		new XMLSaveImpl(Map.of(), new XMIHelperImpl(), "UTF-8").save(null, document,
+				Map.of(XMLResource.OPTION_ROOT_OBJECTS, List.of(eObject)), domMapper);
+		return document.getDocumentElement();
+	}
+
+	private static NamespaceContext createNamespaceContext(Element element) {
+		element.setAttributeNS(XMLConstants.XMLNS_ATTRIBUTE_NS_URI, "xmlns:ecore", EcorePackage.eNS_URI);
+		NamedNodeMap attributes = element.getAttributes();
+		Map<String, String> xmlnsPrefixMap = IntStream.range(0, attributes.getLength()).mapToObj(attributes::item)
+				.map(Attr.class::cast).filter(a -> "xmlns".equals(a.getPrefix()))
+				.collect(Collectors.toMap(Attr::getLocalName, Attr::getValue));
+		Map<String, String> xmlnsPrefixMapInverse = new HashMap<>();
+		xmlnsPrefixMap.forEach((prefix, namespace) -> xmlnsPrefixMapInverse.put(namespace, prefix));
+
+		return new NamespaceContext() {
+			@Override
+			public String getNamespaceURI(String prefix) {
+				return xmlnsPrefixMap.get(prefix);
+			}
+
+			@Override
+			public String getPrefix(String namespaceURI) {
+				return xmlnsPrefixMapInverse.get(namespaceURI);
+			}
+
+			@Override
+			public Iterator<String> getPrefixes(String namespaceURI) {
+				String prefix = getPrefix(namespaceURI);
+				return prefix == null ? Collections.emptyIterator() : List.of(prefix).iterator();
+			}
+		};
+	}
+
+	private static class DOMMapping extends DefaultDOMHandlerImpl {
+
+		public Element getElement(Object object) {
+			for (Map.Entry<Node, Object> entry : nodeToObject.entrySet()) {
+				if (Objects.equals(entry.getValue(), object)) {
+					return (Element) entry.getKey();
+				}
+			}
+			return null;
+		}
+	}
+
+	private static Optional<Object> reconstructReferenceList(EObject first, String xpath, Iterator<Object> iterator) {
+		EReference containment = first.eContainmentFeature();
+		if (containment != null && containment.getUpperBound() == ETypedElement.UNBOUNDED_MULTIPLICITY
```

Comment:
Better to use org.eclipse.emf.ecore.ETypedElement.isMany():
```
  public boolean isMany()
  {
    int upper = getUpperBound();
    return upper > 1 || upper == UNBOUNDED_MULTIPLICITY;
  }
```

Reply:
Yes, absolutely. Thanks for this remark. I forgot that this method exists.